src/client/datascience/notebook/notebookControllerManager.ts

```diff
@@ -0,0 +1,274 @@
+import { INotebookProvider } from '../types';
+import { JupyterNotebookView } from './constants';
+import { getNotebookMetadata, isJupyterKernel, isJupyterNotebook, trackKernelInNotebookMetadata } from './helpers/helpers';
+import { VSCodeNotebookController } from './notebookExecutionHandler';
+import { INotebookControllerManager } from './types';
+/**
+ * This class tracks notebook documents that are open and the provides NotebookControllers for
+ * each of them
+ */
+@injectable()
+export class NotebookControllerManager implements INotebookControllerManager, IExtensionSingleActivationService {
+    private controllerMapping = new WeakMap<NotebookDocument, { selected: VSCodeNotebookController | undefined, controllers: VSCodeNotebookController[] }>();
+    private readonly _onNotebookControllerSelected: EventEmitter<{ notebook: NotebookDocument, controller: VSCodeNotebookController }>;
+
+    private isLocalLaunch: boolean;
+    constructor(
+        @inject(IVSCodeNotebook) private readonly notebook: IVSCodeNotebook,
+        @inject(IDisposableRegistry) private readonly disposables: IDisposableRegistry,
+        @inject(IExtensions) private readonly extensions: IExtensions,
+        @inject(ILocalKernelFinder) private readonly localKernelFinder: ILocalKernelFinder,
+        @inject(IConfigurationService) private readonly configuration: IConfigurationService,
+        @inject(INotebookProvider) private readonly notebookProvider: INotebookProvider,
+        @inject(ICommandManager) private readonly commandManager: ICommandManager,
+        @inject(IExtensionContext) private readonly context: IExtensionContext,
+        @inject(IKernelProvider) private readonly kernelProvider: IKernelProvider,
+        @inject(PreferredRemoteKernelIdProvider)
+        private readonly preferredRemoteKernelIdProvider: PreferredRemoteKernelIdProvider,
+        @inject(IRemoteKernelFinder) private readonly remoteKernelFinder: IRemoteKernelFinder,
+        @inject(INotebookStorageProvider) private readonly storageProvider: INotebookStorageProvider,
+    ) {
+        this._onNotebookControllerSelected = new EventEmitter<{ notebook: NotebookDocument, controller: VSCodeNotebookController }>();
+        this.isLocalLaunch = isLocalLaunch(this.configuration);
+    }
+
+    get onNotebookControllerSelected() {
+        return this._onNotebookControllerSelected.event;
+    }
+
+    public async activate(): Promise<void> {
+        // Sign up for document either opening or closing
+        this.notebook.onDidOpenNotebookDocument(this.onDidOpenNotebookDocument, this, this.disposables);
+        this.notebook.onDidCloseNotebookDocument(this.onDidCloseNotebookDocument, this, this.disposables);
+
+        // Be aware of if we need to re-look for kernels on extension change
+        this.extensions.onDidChange(this.onDidChangeExtensions, this, this.disposables);
+    }
+
+    private onDidChangeExtensions() {
+        // IANHU: Need to invalidate kernels here?
```

Comment:
Curious what were we planning on putting here?
Is this for if user installs Python extenison?

Reply:
Yeah, I believe that's what it was for before. Back in the old pull provider we would signal onDidChangeKernels when extensions changed. Now we need to instead dispose and regenerate NotebookControllers. Just saving this with a todo: as I think we can do it after most everything else is working (i.e. not a super important scenario, you can just close and open the file to get a new list).